_posts/2015-05-11-traits.md

```diff
@@ -0,0 +1,420 @@
+
+### The future
+
+One of the primary ways that languages tend to evolve is in their abstraction
+facilities, and Rust is no exception: many of our [post-1.0 priorities][post1]
+are extensions of the trait system in one direction or another. Here are some
+highlights.
+
+* **Statically dispatched outputs**. Right now, it's possible for functions to
+  use generics for their parameters, but there's no equivalent for their
+  results: you cannot say "this function returns a value of some type that
+  implements the `Iterator` trait" and have that abstraction compiled away.
+  This is particularly problematic when you want to return a closure that you'd
+  like to be statically-dispatched -- you simply can't, in today's Rust. We want
+  to make this possible, and [have some ideas already][impl-trait].
+
+* **Specialization**. Rust does not allow overlap between trait implementations,
+  so there is never ambiguity about which code to run. On the other hand, there
+  are some cases where you can give a "blanket" implementation for a wide range
+  of types, but would then like to provide a more specialized implementation for
+  a few cases, often for performance reasons. We hope to propose a design in the
+  near future.
+
+* **Higher-kinded types** (HKT). Traits today can only be applied to *types*,
+  not *type constructors* -- that is, to things like `Vec<u8>`, not to `Vec`
+  itself. This limitation makes it difficult to provide a good set of container
+  traits, which are therefore not included in the current standard library. HKT
+  is a major, cross-cutting feature that will represent a big step forward in
+  Rust's abstraction capabilities.
+
+* **Efficient re-use**. Finally, while traits provide some mechanisms for
+  reusing code (which we didn't cover above), there are still some patterns of
+  reuse that don't fit well into the language today -- notably, object-oriented
+  hierarchies found in things like the DOM, GUI frameworks, and many
+  games. Accommodating these use cases without adding too much overlap or
+  complexity is a very interesting design problem, and one that Niko Matsakis
+  has started a separate [blog series][virtual] about. It's not yet clear
+  whether this can all be done with traits, or whether some other ingredients
+  are needed.
+
+Of course, we're at the eve of the 1.0 release, and it will take some time for
+the dust to settle, and for the community to have enough experience to start
+landing these extensions. But that makes it an exciting time to get involved:
+from influencing the design at this early stage, to working on implementation,
+to trying out different use cases in your own code -- we'd love to have your
+help!
+
+[zero-cost-cpp]: http://www.stroustrup.com/abstraction-and-machine.pdf
+[fearless]: http://blog.rust-lang.org/2015/04/10/Fearless-Concurrency.html
+[dst5]: http://smallcultfollowing.com/babysteps/blog/2014/01/05/dst-take-5/
+[adhoc]: dl.acm.org/citation.cfm?id=75283
```

Comment:
I think this link needs a leading `http` in front (right now the link is broken)